### PR TITLE
Check if cheatsheets are up to date on CI

### LINKED ISSUE #370: Check cheatsheet is up to date in CI
**Is your feature request related to a problem? Please describe.**
In lazygit we have a cheatsheet [check script](https://github.com/jesseduffield/lazygit/tree/master/pkg/cheatsheet) which tells us whether the cheatsheet we currently have is in line with what would be generated by the cheatsheet generation script.

We want to port this over to lazydocker and also run it on CI (see the lazygit approach for [comparison](https://github.com/jesseduffield/lazygit/blob/master/.github/workflows/ci.yml#L161))

This will ensure that nobody accidentally makes a change while forgetting to update the cheatsheet

### PR DESCRIPTION
**Pull request by gusandrioli**
This PR resolves #370 

# Changes
- Moved cheatsheet generate script to `pkg/cheatsheet/generate.go`.
- Added `pkg/cheatsheet/check.go` to validate if cheatsheets are up to date.
- Added GHA step to validate if cheatsheets are up to date on `ci.yaml`.

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -408,3 +411,31 @@ func IsValidHexValue(v string) bool {
...
+			panic(err)
+		}
+
+		path = filepath.Dir(path)
+
+		if path == "/" {
+			log.Fatal("must run in lazydocker folder or child folder")
+		}
+	}
+}
```
Comment (gusandrioli) [inline pkg/utils/utils.go:441]: I copied over from `lazygit`, but I was wondering if a shared package would fit this case (Maybe `lazyutils`?). There seems to be quite a bit of overlap between them. Even these scripts for cheatsheets could be used in both projects. Let me know your thoughts 😄.

File: pkg/cheatsheet/validate.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			ToDate:   "",
+			Context:  1,
+		}); err != nil {
+			log.Fatalf("Error occurred while checking if cheatsheets are up to date: %v", err)
+		}
+		fmt.Printf(
+			"\nCheatsheets are out of date. Please run `%s` at the project root and commit the changes. "+
+				"If you run the script and no keybindings files are updated as a result, try rebasing onto master"+
+				"and trying again.\n",
+			generateCheatsheetCmd,
```
Comment (gusandrioli) [inline pkg/cheatsheet/validate.go:54]: Updated to use the const.

File: pkg/cheatsheet/generate.go [hunk 3]
```diff
@@ -33,13 +41,20 @@ func main() {
...
 		}
 
 		bindingSections := getBindingSections(mApp)
 		content := formatSections(mApp, bindingSections)
+		content = fmt.Sprintf(
+			"_This file is auto-generated. To update, make the changes in the "+
+				"pkg/i18n directory and then run `%s` from the project root._\n\n%s",
+			generateCheatsheetCmd,
+			content,
+		)
```
Comment (gusandrioli) [inline pkg/cheatsheet/generate.go:57]: I copied over these changes from #373 as i needed the constants as well. If this gets approved/merged first we can close #373 .

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -35,6 +35,9 @@ jobs:
           key: ${{runner.os}}-go-${{hashFiles('**/go.sum')}}-test
           restore-keys: |
             ${{runner.os}}-go-
+      - name: Check Cheatsheet
+        run: |
+          go run scripts/cheatsheet/main.go check
```
Comment (gusandrioli) [inline .github/workflows/ci.yml:40]: Very similar to `lazygit`'s . Still left it inside `CI` job as it seems fit the context. Let me know if you think i should separate the jobs.

**Comment by gusandrioli** [comment]
@jesseduffield This one is ready for review as well!

File: pkg/utils/utils.go (hunk 1 above)
Comment (jesseduffield) [reply pkg/utils/utils.go:441]: @gusandrioli I actually made a repo for this exact purpose over the weekend: https://github.com/jesseduffield/lazycore. If you would like to raise a PR on that repo for shared cheatsheet functionality, I'll happily review :) Also happy to give pointers

Comment (gusandrioli) [reply pkg/utils/utils.go:441]: Ah neat! I'll open a PR for cheatsheet functionality there then 😄.

**Comment by gusandrioli** [comment]
@jesseduffield Added a PR on lazycore: https://github.com/jesseduffield/lazycore/pull/3. Also, investigating why the failure on `windows-latest ci`. I saw `lazygit` doesn't have this problem, so let me know if this is a known issue.

**Comment by jesseduffield** [comment]
@gusandrioli nice, I've merged that PR. There's a `scripts/bump_lazycore.sh` script you can run to bump it for you in this repo.

I'm not sure why windows-latest ci would be failing. I am aware of it being slow but not failing

**Comment by jesseduffield** [comment]
@gusandrioli actually I bet it's because windows uses different line ending characters than unix. I'm not sure how that's possible when the script is only using newlines though

**Comment by gusandrioli** [comment]
@jesseduffield No luck yet here. Seems like on `lazygit` it only runs `check-codebase` on ubuntu machines (and `Check Cheatsheet` within there). We'd probably have the same issue there. I'll probably separate the jobs here to match that.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -66,6 +66,34 @@ jobs:
...
+          key: ${{runner.os}}-go-${{hashFiles('**/go.sum')}}-build
+          restore-keys: |
+            ${{runner.os}}-go-
+      - name: Check Cheatsheet
+        run: |
+          go run scripts/cheatsheet/main.go check
+      - name: Check Vendor Directory
+        # ensure our vendor directory matches up with our go modules
+        run: |
+          go mod vendor && git diff --exit-code || (echo "Unexpected change to vendor directory. Run 'go mod vendor' locally and commit the changes" && exit 1)
```
Comment (gusandrioli) [inline .github/workflows/ci.yml:96]: Now we check for cheatsheet diffs here and ensure our vendor dir matches the modules (same as lazygit).

File: vendor/modules.txt [hunk 6]
```diff
@@ -100,9 +100,6 @@ github.com/go-errors/errors
 # github.com/gogo/protobuf v1.3.1
 ## explicit
 github.com/gogo/protobuf/proto
-# github.com/golang-collections/collections v0.0.0-20130729185459-604e922904d3
-## explicit
-github.com/golang-collections/collections/stack
```
Comment (gusandrioli) [inline vendor/modules.txt:105]: Ran `go mod vendor` as now there's validation on CI.

**Review by gusandrioli — COMMENTED** [review]
@jesseduffield CI passing now :)

**Comment by jesseduffield** [comment]
Amazing work @gusandrioli ! Merging :)

**Pull request merged**